features/dd-sdk-android-trace/src/main/java/com/datadog/opentracing/PendingTrace.java

```diff
@@ -25,410 +27,469 @@
+                        final Iterator<DDSpan> it = iterator();
+                        while (it.hasNext()) {
+                            final DDSpan span = it.next();
+                            if (span != rootSpan) {
+                                partialTrace.add(span);
+                                completedSpanCount.decrementAndGet();
+                                it.remove();
+                            }
+                        }
+                        tracer.write(partialTrace);
+                    }
+                }
             }
-            tracer.write(partialTrace);
-          }
         }
-      }
     }
-  }
-
-  private synchronized void write() {
-    if (isWritten.compareAndSet(false, true)) {
-      removePendingTrace();
-      if (!isEmpty()) {
-        tracer.write(this);
-      }
-    } else {
-      internalLogger.log(
-          InternalLogger.Level.ERROR,
-          InternalLogger.Target.USER,
-          () -> "Trace " + traceId + " write ignored: isWritten already true",
-          null,
-          false,
-          new HashMap<>()
-      );
+
+    private synchronized void write() {
+        if (isWritten.compareAndSet(false, true)) {
+            removePendingTrace();
+            if (!isEmpty()) {
+                tracer.write(this);
+            }
+        } else {
+            internalLogger.log(
+                    InternalLogger.Level.ERROR,
+                    InternalLogger.Target.USER,
+                    () -> "Trace " + traceId + " write ignored: isWritten already true",
+                    null,
+                    false,
+                    new HashMap<>()
+            );
+        }
     }
-  }
-
-  public synchronized boolean clean() {
-    Reference ref;
-    int count = 0;
-    while ((ref = referenceQueue.poll()) != null) {
-      weakReferences.remove(ref);
-      if (isWritten.compareAndSet(false, true)) {
-        removePendingTrace();
-        // preserve throughput count.
-        // Don't report the trace because the data comes from buggy uses of the api and is suspect.
-        tracer.incrementTraceCount();
-      }
-      count++;
-      expireReference();
+
+    public synchronized boolean clean() {
+        Reference ref;
+        int count = 0;
+        while ((ref = referenceQueue.poll()) != null) {
+            weakReferences.remove(ref);
+            if (isWritten.compareAndSet(false, true)) {
+                removePendingTrace();
+                // preserve throughput count.
+                // Don't report the trace because the data comes from buggy uses of the api and is suspect.
+                tracer.incrementTraceCount();
+            }
+            count++;
+            expireReference();
+        }
+        if (count > 0) {
+            // TODO attempt to flatten and report if top level spans are finished. (for accurate metrics)
+        } else if (pendingReferenceCount.get() == 0) {
```

Comment:
I'm not quite sure about this change: let's say we created a new `PendingTrace` without any spans inside and we add spans only _after_ first cycle of the `SPAN_CLEANER` is completed - with this change `removePendingTrace` will be called even before we manage to register first span, no? 🤔